### ISSUE TITLE
[Feature Request]: Keep links when exporting to pdf

### ISSUE THREAD

**Issue by ForeverRainbow** [giuspen/cherrytree#942]
Hi.
Currently exporting to pdf will result in links being lost, both internal (i.e to other nodes -> exported pages) and external (e.g to a website). 

It would be really useful if links could be exported, since especially when navigating lots of notes, not having my "more info here" links work is quite a big issue. I can always just read these notes in cherrytree of course but its very helpful to be able to distribute them in a more or less universal format like pdf.

Internal and external links are supported in pdf afaik, but I have little to no experience with pdf beyond just using it. I am happy to try and implement this if it is doable though.

**Comment by ForeverRainbow** [comment]
Just to say as well, `Pandoc` does support pdf conversion so that might be an option just for external links, however if internal links are to be kept as well I think something more complex is needed

**Comment by giuspen** [comment]
I was hoping that with Gtk3 or at least Gtk4 the pdf would support links... This feature is quite high in the number of requests so definitely a valid feature to investigate, thanks!

**Comment by txe** [comment]
Cairo supports links
https://www.cairographics.org/manual/cairo-Tags-and-Links.html
https://gitlab.freedesktop.org/cairo/cairo/-/issues/357

**Comment by ForeverRainbow** [comment]
@txe So it would require a Cario dependency? Or does cherrytree already use Cario? Looking at the code for `CtExport2Pdf` it looks like it just translates to pango and then mysteriously a pdf appears (I am guess gtk can create pdfs but cannot find any documentation on it)

**Comment by txe** [comment]
Agree, this stuff is confusing. Both are already used. Cairo is low-level render. Pango is high-level render and uses Cairo.
https://en.wikipedia.org/wiki/Pango

Things are go on in
`CtPrint::_on_begin_print_text`   - to calculate text/images/other positions (layout)
`CtPrint::_on_draw_page_tex`t - is actual drawing

**Comment by ForeverRainbow** [comment]
@txe It looks like the "pipeline" for pdf creation goes something like: `text buffer -> pango -> adjust output positions -> output pango using cairo`. Which means that after the text buffer stage anything not supported in pango markup (i.e links) is lost (with the exception of widgets). 

I am not sure how to get around this without redesigning the pipeline for pdf exporting which would be pretty massive and (maybe?) extreme. Is there anything I am missing here with the export process?

**Comment by txe** [comment]
Yeah, it doesn't look simple.
But if I were you, I firstly tried to draw any random link in some place. Aftre that it would be easier to figure out how to pass necessary data.

**Comment by ForeverRainbow** [comment]
I got test links working with `cairo_open_tag` and all that in `_on_draw_page_text`, the problem is keeping the information that x is a link, which doesn't seem to be possible currently as pango has no `link` tag or similar (and I don't think its a good idea to try and preprocess the pango text)

**Comment by txe** [comment]
> (and I don't think its a good idea to try and preprocess the pango text)

it is about putting a fake tag `<link>` into pango text and then cutting it out in `_on_begin_print_text`?

**Comment by ForeverRainbow** [comment]
Thats what I was considering but there is quick a large margin for error with that (what if the text is processed before reaching the check stage), if its the only way without redesigning the exporter though I guess its acceptable

**Comment by txe** [comment]
CtPango is used only to prepare data for pdf and changing it is not a big deal. But it a lot of work to get link from text and then some complicated data structure in pass it and then to change a lot of code to process it.

I thinks simpler is better. I would put a link between `<link>` tags in pango serializer, then in `CtPrint::print_text` cut it out (any xml reader) and put link in CtPrintLinkProxy. After that things start to be much simpler.

FIY, there is a dependence between `pango_text` (std::vector<std::string>) and widgets. They actually create layers, and text is always is the first and the last ones
```
pango_text[0]
widget[0]
pango_text[1]
widget[1]
pango_text[2]
``` 
Then you cut text, you need to figure out the right index of new widget

**Comment by txe** [comment]
Or pango data can be redesigned. instead of having list of pango text and list of widgets, it can be one list of `CtPangoElement` which can be either text or widget or link.
If you want to do this way, I suggest to PR to introduce this redesign, and only then to start on adding a link element.

**Comment by ForeverRainbow** [comment]
The redesign would be a cleaner way I think

**Comment by txe** [comment]
OK

**Comment by ForeverRainbow** [comment]
I think it may actually be better to just add some custom parsing to the pango text, I have been trying to redesign this and it is really quite painful.

**Comment by txe** [comment]
OK

**Comment by ForeverRainbow** [comment]
@txe I am going to go ahead with implementing the links now, is there anything major you want to rework which might conflict?

**Comment by txe** [comment]
@ForeverRainbow , no, nothing like that.

**Comment by ForeverRainbow** [comment]
Should I close this? I didn't mark it as resolved in the pull request because it only applies to the C++ version

**Comment by txe** [comment]
Thanks @ForeverRainbow , C++ version is enough. Also, I closed #514
